File: editorial/chap04/python/answer_B24.py
import bisect

# 配列 A の最長増加部分列（LIS）の長さを計算する
# 配列 dp を使わない実装方法を利用している
def Get_LISvalue(A):
	LEN = 0
	L = []
	for i in range(N):
		pos = bisect.bisect_left(L, A[i])
		if pos == LEN:
			L.append(A[i])
			LEN += 1
		else:
			L[pos] = A[i]
	return LEN

# 入力
N = int(input())
X = [ None ] * N
Y = [ None ] * N
for i in range(N):
	X[i], Y[i] = map(int, input().split())

# ソート
tmp = []
for i in range(N):
	tmp.append([X[i], -Y[i]])
tmp.sort()

# LIS を求めるべき配列は？
A = []
for i in range(N):
	A.append(-tmp[i][1])

# 出力
print(Get_LISvalue(A))

File: editorial/chap04/answer_B24.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
 
int N, X[100009], Y[100009];
int LEN, L[100009];
 
// 配列 A の最長増加部分列（LIS）の長さを計算する
// 配列 dp を使わない実装方法を利用している
int Get_LISvalue(vector<int> A) {
	LEN = 0;
	for (int i = 1; i <= A.size(); i++) L[i] = 0;
 
	// 動的計画法
	for (int i = 0; i < A.size(); i++) {
		int pos = lower_bound(L + 1, L + LEN + 1, A[i]) - L;
		L[pos] = A[i];
		if (pos > LEN) LEN += 1;
	}
	return LEN;
}
 
int main() {
	// 入力
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> X[i] >> Y[i];
 
	// ソート
	vector<pair<int, int>> tmp;
	for (int i = 1; i <= N; i++) tmp.push_back(make_pair(X[i], -Y[i]));
	sort(tmp.begin(), tmp.end());
	
	// LIS を求めるべき配列は？
	vector<int> A;
	for (int i = 0; i < tmp.size(); i++) {
		A.push_back(-tmp[i].second);
	}
 
	// 出力
	cout << Get_LISvalue(A) << endl;
	return 0;
}
